design-system/scripts/generate-design-tokens.js

```diff
@@ -21,14 +21,99 @@ const endProgram = (message) => {
   process.exit(1);
 };
 
-const TOKEN_REGEX =
-  /^(\w+(?:-\w+)(?:-\w+)?)(?:-for-(\w+(?:-\w+)?))?(?:-when-([\w-]+?))?(?:-on-([\w-]+?))?$/i;
+const ALLOWED_CSS_VALUES_IN_CHOICES = /px|none|hsla/;
+
+const isAllowedCssChoice = (choice) =>
+  choice.match(ALLOWED_CSS_VALUES_IN_CHOICES) !== null;
 
 const supportedStates = Object.keys(definitions.states);
 const supportedComponentGroups = Object.keys(definitions.componentGroups);
+const supportedVariants = Object.keys(definitions.variants || {});
 
 const designTokens = {};
 
+/*
+  Allowed patterns with examples:
+    - <attribute>-for-<component-group>
+      + background-color-for-tag
+    - <attribute>-for-<component-group>-when-<state>-
+      + background-color-for-button-when-disabled
+    - <attribute>-for-<component-group>-as-<variant>-
+      + border-radius-for-button-as-big
+    - <attribute>-for-<component-group>-as-<variant>-when-<state>
+    + border-for-button-as-secondary-when-hovered
+    - <attribute>-for-<component-group>-as-<variant>-as-<variant>-
+    + border-radius-for-button-as-icon-as-small
+    - <attribute>-for-<component-group>-as-<variant>-as-<variant>-when-<state>-
+      + border-radius-for-button-as-icon-as-small-when-disabled
+*/
+function parseToken(token) {
+  const parts = token.split('-');
+  let partType = 'cssRule';
+  let newVariant = false;
+  return parts.reduce(
+    (agg, part) => {
+      if (['for', 'as', 'when'].includes(part)) {
+        partType = part;
+        if (part === 'as') newVariant = true;
+        return agg;
+      }
+
+      if (partType === 'for') {
+        agg.componentGroup = `${agg.componentGroup}${
+          !!agg.componentGroup ? '-' : ''
+        }${part}`;
```

Comment:
Since this pattern is used in a few places we could have an external function:
```js
const joinParts = (preceding, following) =>
  `${preceding}${preceding.length > 0 ? '-' : ''}${following}`;
```
and use it like
```js
agg.componentGroup = joinParts(agg.componentGroup, part);
```

Reply:
cleaner 👌 

Updated in b008efe0cce4e6cffd60e61ff554e5480a400525